{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:kipoi_utils.external.related.mixins:Unrecognized fields for DataLoaderDescription: {'postprocessing'}. Available fields are {'output_schema', 'type', 'info', 'dependencies', 'defined_as', 'args', 'writers', 'path'}\n",
      "WARNING:kipoi_utils.external.related.mixins:Unrecognized fields for DataLoaderDescription: {'postprocessing'}. Available fields are {'output_schema', 'type', 'info', 'dependencies', 'defined_as', 'args', 'writers', 'path'}\n",
      "WARNING:kipoi_utils.external.related.mixins:Unrecognized fields for DataLoaderDescription: {'postprocessing'}. Available fields are {'output_schema', 'type', 'info', 'dependencies', 'defined_as', 'args', 'writers', 'path'}\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import warnings\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import matplotlib.pyplot as plt\n",
    "sys.path.append(\"/home/wangrr/Analysis/epiRNA/src\")\n",
    "import epiRNA as erna\n",
    "\n",
    "workdir = \"/data/wangrr/Analysis/epiRNA/workdir2/benchmark\"\n",
    "os.makedirs(workdir, exist_ok=True); os.chdir(workdir)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Load models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# model config:\n",
    "config = \"/home/wangrr/Analysis/epiRNA/src/experiments/model_v2/configs/basic_model.yaml\"\n",
    "\n",
    "# model checkpoint:\n",
    "ckpt_dir = \"/data/wangrr/Analysis/epiRNA/models\"\n",
    "checkpoints = {\n",
    "    \"complete\": f\"{ckpt_dir}/basic_fold/version_0/checkpoints/epoch=18-step=215327-val_ePCC_mixed=0.7823.ckpt\",\n",
    "    \"maskrna\": f\"{ckpt_dir}/v2maskrna/version_0/checkpoints/epoch=11-step=135996-val_ePCC_mixed=0.4895.ckpt\",\n",
    "    \"maskseq\": f\"{ckpt_dir}/v2maskseq/version_0/checkpoints/epoch=02-step=33999-val_ePCC_mixed=0.4829.ckpt\",\n",
    "    \"maskrbp\": f\"{ckpt_dir}/v2maskrbp/version_0/checkpoints/epoch=10-step=124663-val_ePCC_mixed=0.6686.ckpt\"\n",
    "}\n",
    "\n",
    "# load models\n",
    "model_zoo = {}\n",
    "for key, ckpt in checkpoints.items():\n",
    "    model = erna.evals.load_model(ckpt, config)\n",
    "    model.eval()\n",
    "    model_zoo[key] = model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Load samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_version', 'gene_name', 'gene_source', 'gene_biotype', 'havana_gene', 'havana_gene_version', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'havana_transcript', 'havana_transcript_version', 'tag', 'transcript_support_level', 'exon_number', 'exon_id', 'exon_version', 'ccds_id', 'protein_id', 'protein_version']\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_version', 'gene_name', 'gene_source', 'gene_biotype', 'havana_gene', 'havana_gene_version', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'havana_transcript', 'havana_transcript_version', 'tag', 'transcript_support_level', 'exon_number', 'exon_id', 'exon_version', 'ccds_id', 'protein_id', 'protein_version']\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "\n",
    "# load mock samples\n",
    "datainfo = pd.read_excel(\"/home/wangrr/Analysis/epiRNA/data/mRNA_multiomic_sample_info.xlsx\", sheet_name=\"m6A\")\n",
    "\n",
    "# pick human and mouse m6A samples\n",
    "# m6A_train_samples = [\"Adipose-1\", \"Aorta-1\", \"Appendix-1\", \"Esophagus-1\", \"GOS-1\", \"Hela-1\", \"HT29-1\", \"Hypothalamus-1\", \"Jejunum-1\", \"Liver-1\", \"Prostate-1\", \"Skin-1\", \"Testis-1\", \"Tongue-1\", \"Trachea-1\", \"WPMY-1\", \"MT4-1\", \"GSC11-1\", \"iSLK-1\", \"TIME-1\", \"mHeart-1\", \"mSpleen-1\", \"mLung-1\", \"mLiver-1\", \"mCerebrum-1\", \"mCerebellum-1\", \"mBrainstem-1\", \"mHypothalamus-1\", \"mBMDC-1\", \"mEF-1\", \"mNPC-1\", \"mESC-1\", \"mStriatum-1\", \"mPeritoneal_macrophages-1\", \"mB16-OVA-1\", \"miPSC-1\", \"mNSC-1\", \"mForebrain-1\", \"mKidney-1\", \"mHypothalamus-2\"]\n",
    "m6A_train_samples = [\"Adipose-1\"]\n",
    "\n",
    "datainfo = datainfo[datainfo[\"SID\"].isin(m6A_train_samples)].reset_index(drop=True)\n",
    "\n",
    "ref_dir = \"/data/wangrr/Analysis/epiRNA/Dataset/Reference\"\n",
    "genome = {\n",
    "    \"human\":{\n",
    "        \"fasta\": f\"{ref_dir}/Homo_sapiens.GRCh38.dna.primary_assembly.84.fa\",\n",
    "        \"gtf\": f\"{ref_dir}/Homo_sapiens.GRCh38.84.gtf\"\n",
    "    },\n",
    "    \"mouse\":{\n",
    "        \"fasta\": f\"{ref_dir}//Mus_musculus.GRCm38.dna.primary_assembly.84.fa\",\n",
    "        \"gtf\": f\"{ref_dir}/Mus_musculus.GRCm38.84.gtf\"\n",
    "    }\n",
    "}\n",
    "\n",
    "human_gtf_df = erna.pp.GTFLoader(gtf_path=genome[\"human\"][\"gtf\"], zero_based=True)\n",
    "mouse_gtf_df = erna.pp.GTFLoader(gtf_path=genome[\"mouse\"][\"gtf\"], zero_based=True)\n",
    "\n",
    "chrom_kwargs = {\n",
    "    'human': {\n",
    "        'train_chroms': [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"9\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\", \"20\", \"21\", \"22\"],\n",
    "        'val_chroms': [\"8\"],\n",
    "        'test_chroms': [\"10\"]\n",
    "    },\n",
    "    'mouse': {\n",
    "        'train_chroms': [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"9\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\"],\n",
    "        'val_chroms': [\"8\"],\n",
    "        'test_chroms': [\"10\"]\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"/home/wangrr/Analysis/epiRNA/data/rbp_expression_dict.pkl\", \"rb\") as f:\n",
    "    rbp_dict = pickle.load(f)['rbp_dict']\n",
    "    \n",
    "with open(\"/home/wangrr/Analysis/epiRNA/data/rbp_expression_dict_zero.pkl\", \"rb\") as f:\n",
    "    rbp_dict_zero = pickle.load(f)['rbp_dict']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "fulldataset_kwargs = {\n",
    "    'human_fasta_path': genome[\"human\"][\"fasta\"],\n",
    "    'mouse_fasta_path': genome[\"mouse\"][\"fasta\"],\n",
    "    \"modality_to_index\": {\"m6A_human\": 0, \"m6A_mouse\": 1},\n",
    "    'bigwig_dir': \"/data/wangrr/Analysis/epiRNA/Dataset/processed/4-coverage\",\n",
    "    'datainfo': datainfo,\n",
    "    'rbp_dict': rbp_dict,\n",
    "    'human_gtf_df': human_gtf_df,\n",
    "    'mouse_gtf_df': mouse_gtf_df,\n",
    "    'return_bin_input': True,\n",
    "    'seq_len': 65536,\n",
    "    'mask_type': None\n",
    "}\n",
    "\n",
    "maskrnadataset_kwargs = fulldataset_kwargs.copy()\n",
    "maskrnadataset_kwargs['mask_type'] = 'expression'\n",
    "\n",
    "maskseqdataset_kwargs = fulldataset_kwargs.copy()\n",
    "maskseqdataset_kwargs['mask_type'] = 'sequence'\n",
    "\n",
    "maskrbpdataset_kwargs = fulldataset_kwargs.copy()\n",
    "maskrbpdataset_kwargs['rbp_dict'] = rbp_dict_zero\n",
    "\n",
    "fulldatamodule = erna.ds.GeneDataModule(\n",
    "    dataset_cls=erna.ds.GeneDataset,\n",
    "    dataset_kwargs=fulldataset_kwargs,\n",
    "    chrom_kwargs=chrom_kwargs,\n",
    "    batch_size=32,\n",
    "    num_workers=8\n",
    ")\n",
    "\n",
    "maskrnadatamodule = erna.ds.GeneDataModule(\n",
    "    dataset_cls=erna.ds.GeneDataset,\n",
    "    dataset_kwargs=maskrnadataset_kwargs,\n",
    "    chrom_kwargs=chrom_kwargs,\n",
    "    batch_size=32,\n",
    "    num_workers=8\n",
    ")\n",
    "\n",
    "maskseqdatamodule = erna.ds.GeneDataModule(\n",
    "    dataset_cls=erna.ds.GeneDataset,\n",
    "    dataset_kwargs=maskseqdataset_kwargs,\n",
    "    chrom_kwargs=chrom_kwargs,\n",
    "    batch_size=32,\n",
    "    num_workers=8\n",
    ")\n",
    "\n",
    "maskrbpdatamodule = erna.ds.GeneDataModule(\n",
    "    dataset_cls=erna.ds.GeneDataset,\n",
    "    dataset_kwargs=maskrbpdataset_kwargs,\n",
    "    chrom_kwargs=chrom_kwargs,\n",
    "    batch_size=32,\n",
    "    num_workers=8\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total samples: 17509\n",
      "Omics types: ['m6A']\n",
      "Total samples: 669\n",
      "Omics types: ['m6A']\n",
      "Total samples: 729\n",
      "Omics types: ['m6A']\n"
     ]
    }
   ],
   "source": [
    "fulldatamodule.setup(stage='test')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Start benchmarking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total samples: 17509\n",
      "Omics types: ['m6A']\n",
      "Total samples: 669\n",
      "Omics types: ['m6A']\n",
      "Total samples: 729\n",
      "Omics types: ['m6A']\n",
      "Evaluating Model: complete on Data: complete during train phase\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Benchmarking PCC:   0%|          | 0/547 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Benchmarking PCC: 100%|██████████| 547/547 [02:06<00:00,  4.34it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating Model: complete on Data: complete during val phase\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Benchmarking PCC: 100%|██████████| 21/21 [00:07<00:00,  2.86it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating Model: complete on Data: complete during test phase\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Benchmarking PCC:  13%|█▎        | 3/23 [00:02<00:13,  1.47it/s]Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7f9f5c8b8a40>\n",
      "Traceback (most recent call last):\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1604, in __del__\n",
      "    self._shutdown_workers()\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1587, in _shutdown_workers\n",
      "    if w.is_alive():\n",
      "       ^^^^^^^^^^^^\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/multiprocessing/process.py\", line 160, in is_alive\n",
      "    assert self._parent_pid == os.getpid(), 'can only test a child process'\n",
      "AssertionError: can only test a child process\n",
      "Exception ignored in: Exception ignored in: Exception ignored in: Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7f9f5c8b8a40><function _MultiProcessingDataLoaderIter.__del__ at 0x7f9f5c8b8a40><function _MultiProcessingDataLoaderIter.__del__ at 0x7f9f5c8b8a40>\n",
      "\n",
      "<function _MultiProcessingDataLoaderIter.__del__ at 0x7f9f5c8b8a40>\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1604, in __del__\n",
      "Traceback (most recent call last):\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1604, in __del__\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1604, in __del__\n",
      "      File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1604, in __del__\n",
      "        self._shutdown_workers()self._shutdown_workers()    self._shutdown_workers()\n",
      "\n",
      "self._shutdown_workers()\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1587, in _shutdown_workers\n",
      "\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1587, in _shutdown_workers\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1587, in _shutdown_workers\n",
      "      File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1587, in _shutdown_workers\n",
      "    if w.is_alive():        if w.is_alive():\n",
      "if w.is_alive():\n",
      "if w.is_alive(): \n",
      " \n",
      "       Exception ignored in:     <function _MultiProcessingDataLoaderIter.__del__ at 0x7f9f5c8b8a40> Exception ignored in: Exception ignored in:     \n",
      "<function _MultiProcessingDataLoaderIter.__del__ at 0x7f9f5c8b8a40><function _MultiProcessingDataLoaderIter.__del__ at 0x7f9f5c8b8a40>    Traceback (most recent call last):\n",
      "\n",
      "\n",
      "      File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1604, in __del__\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      " ^^   File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1604, in __del__\n",
      "    ^  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1604, in __del__\n",
      "^^^self._shutdown_workers()^^^        \n",
      "^^^^self._shutdown_workers()self._shutdown_workers()^  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1587, in _shutdown_workers\n",
      "^^^\n",
      "\n",
      "^^^    ^  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1587, in _shutdown_workers\n",
      "^  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1587, in _shutdown_workers\n",
      "^^if w.is_alive():^^        ^\n",
      "^^^if w.is_alive():if w.is_alive(): ^^^^\n",
      "\n",
      " ^^^^ Exception ignored in:   ^^^^ <function _MultiProcessingDataLoaderIter.__del__ at 0x7f9f5c8b8a40>  ^^^ ^\n",
      "  \n",
      "\n",
      "\n",
      " Traceback (most recent call last):\n",
      "^    File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/multiprocessing/process.py\", line 160, in is_alive\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/multiprocessing/process.py\", line 160, in is_alive\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/multiprocessing/process.py\", line 160, in is_alive\n",
      "   File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1604, in __del__\n",
      "^               \n",
      " ^    assert self._parent_pid == os.getpid(), 'can only test a child process'assert self._parent_pid == os.getpid(), 'can only test a child process'assert self._parent_pid == os.getpid(), 'can only test a child process'   File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/multiprocessing/process.py\", line 160, in is_alive\n",
      "^ self._shutdown_workers()\n",
      "\n",
      "\n",
      "^^    ^AssertionError\n",
      "AssertionErrorAssertionError^assert self._parent_pid == os.getpid(), 'can only test a child process'^^:   File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1587, in _shutdown_workers\n",
      ": ^: ^\n",
      "^can only test a child processcan only test a child process    ^can only test a child process^^AssertionError\n",
      "\n",
      "if w.is_alive():^^\n",
      "^: \n",
      "^^^can only test a child process ^^^ \n",
      "^^^ ^^^ ^^^ ^^ \n",
      "^^   File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/multiprocessing/process.py\", line 160, in is_alive\n",
      "\n",
      "\n",
      "^      File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/multiprocessing/process.py\", line 160, in is_alive\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/multiprocessing/process.py\", line 160, in is_alive\n",
      "^assert self._parent_pid == os.getpid(), 'can only test a child process'        ^\n",
      "assert self._parent_pid == os.getpid(), 'can only test a child process'assert self._parent_pid == os.getpid(), 'can only test a child process'^AssertionError\n",
      "\n",
      "^: AssertionErrorAssertionError^can only test a child process: : ^\n",
      "can only test a child processcan only test a child process^\n",
      "\n",
      "^^^^\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/multiprocessing/process.py\", line 160, in is_alive\n",
      "    assert self._parent_pid == os.getpid(), 'can only test a child process'\n",
      "AssertionError: can only test a child process\n",
      "Exception ignored in: Exception ignored in: Exception ignored in: Exception ignored in: <function _MultiProcessingDataLoaderIter.__del__ at 0x7f9f5c8b8a40><function _MultiProcessingDataLoaderIter.__del__ at 0x7f9f5c8b8a40><function _MultiProcessingDataLoaderIter.__del__ at 0x7f9f5c8b8a40><function _MultiProcessingDataLoaderIter.__del__ at 0x7f9f5c8b8a40>\n",
      "\n",
      "\n",
      "\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1604, in __del__\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1604, in __del__\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1604, in __del__\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1604, in __del__\n",
      "                self._shutdown_workers()self._shutdown_workers()self._shutdown_workers()self._shutdown_workers()\n",
      "\n",
      "\n",
      "\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1587, in _shutdown_workers\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1587, in _shutdown_workers\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1587, in _shutdown_workers\n",
      "  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1587, in _shutdown_workers\n",
      "                if w.is_alive():if w.is_alive():if w.is_alive():if w.is_alive():\n",
      "\n",
      "\n",
      "\n",
      "Exception ignored in:      <function _MultiProcessingDataLoaderIter.__del__ at 0x7f9f5c8b8a40>    \n",
      "Exception ignored in:     Traceback (most recent call last):\n",
      " <function _MultiProcessingDataLoaderIter.__del__ at 0x7f9f5c8b8a40>     File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1604, in __del__\n",
      " \n",
      "    Traceback (most recent call last):\n",
      "          File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1604, in __del__\n",
      "self._shutdown_workers()Exception ignored in:  ^\n",
      "^     <function _MultiProcessingDataLoaderIter.__del__ at 0x7f9f5c8b8a40>^  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1587, in _shutdown_workers\n",
      "^^^self._shutdown_workers()\n",
      "^    ^^\n",
      "^^Traceback (most recent call last):\n",
      "if w.is_alive():^  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1587, in _shutdown_workers\n",
      "^  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1604, in __del__\n",
      "^^^\n",
      "^        ^^ ^^if w.is_alive():self._shutdown_workers()^ ^^\n",
      "\n",
      "^^ ^^   File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/site-packages/torch/utils/data/dataloader.py\", line 1587, in _shutdown_workers\n",
      "^^ ^ ^    ^ ^^ if w.is_alive():^^ ^ ^\n",
      "^^ ^ ^ \n",
      "^^\n",
      " ^   File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/multiprocessing/process.py\", line 160, in is_alive\n",
      "^^  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/multiprocessing/process.py\", line 160, in is_alive\n",
      " ^     ^^^\n",
      "     assert self._parent_pid == os.getpid(), 'can only test a child process'\n",
      "^  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/multiprocessing/process.py\", line 160, in is_alive\n",
      "^assert self._parent_pid == os.getpid(), 'can only test a child process'   File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/multiprocessing/process.py\", line 160, in is_alive\n",
      "^\n",
      "    ^\n",
      "     ^^assert self._parent_pid == os.getpid(), 'can only test a child process'AssertionErrorAssertionError ^assert self._parent_pid == os.getpid(), 'can only test a child process'^\n",
      ": : ^\n",
      "^^AssertionErrorcan only test a child processcan only test a child process^^AssertionError^: \n",
      "^\n",
      "^: ^can only test a child process^^^can only test a child process\n",
      "^^\n",
      "^^\n",
      "^^  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/multiprocessing/process.py\", line 160, in is_alive\n",
      "^^\n",
      "    ^  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/multiprocessing/process.py\", line 160, in is_alive\n",
      "assert self._parent_pid == os.getpid(), 'can only test a child process'^    \n",
      "^assert self._parent_pid == os.getpid(), 'can only test a child process'AssertionError^\n",
      ": \n",
      "AssertionErrorcan only test a child process  File \"/nvme/biosoft/miniforge3/envs/wrr_m6a_env/lib/python3.11/multiprocessing/process.py\", line 160, in is_alive\n",
      ": \n",
      "    can only test a child processassert self._parent_pid == os.getpid(), 'can only test a child process'\n",
      "\n",
      "AssertionError: can only test a child process\n",
      "Benchmarking PCC: 100%|██████████| 23/23 [00:08<00:00,  2.87it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total samples: 17509\n",
      "Omics types: ['m6A']\n",
      "Total samples: 669\n",
      "Omics types: ['m6A']\n",
      "Total samples: 729\n",
      "Omics types: ['m6A']\n",
      "Evaluating Model: complete on Data: maskrna during train phase\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Benchmarking PCC: 100%|██████████| 547/547 [02:08<00:00,  4.26it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating Model: complete on Data: maskrna during val phase\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Benchmarking PCC: 100%|██████████| 21/21 [00:07<00:00,  2.88it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating Model: complete on Data: maskrna during test phase\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Benchmarking PCC: 100%|██████████| 23/23 [00:07<00:00,  2.94it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total samples: 17509\n",
      "Omics types: ['m6A']\n",
      "Total samples: 669\n",
      "Omics types: ['m6A']\n",
      "Total samples: 729\n",
      "Omics types: ['m6A']\n",
      "Evaluating Model: complete on Data: maskseq during train phase\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Benchmarking PCC:  69%|██████▉   | 378/547 [01:29<00:39,  4.24it/s]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[26]\u001b[39m\u001b[32m, line 11\u001b[39m\n\u001b[32m      9\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m phase, dataloader \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m([\u001b[33m\"\u001b[39m\u001b[33mtrain\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mval\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mtest\u001b[39m\u001b[33m\"\u001b[39m], dataloaders):\n\u001b[32m     10\u001b[39m     \u001b[38;5;28mprint\u001b[39m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mEvaluating Model: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmodel_type\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m on Data: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mdata_type\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m during \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mphase\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m phase\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m---> \u001b[39m\u001b[32m11\u001b[39m     summary = \u001b[43merna\u001b[49m\u001b[43m.\u001b[49m\u001b[43mevals\u001b[49m\u001b[43m.\u001b[49m\u001b[43mbenchmark_summary\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m     12\u001b[39m \u001b[43m        \u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m=\u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m     13\u001b[39m \u001b[43m        \u001b[49m\u001b[43mdataloader\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdataloader\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m     14\u001b[39m \u001b[43m        \u001b[49m\u001b[43mn_channels\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m2\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     15\u001b[39m \u001b[43m        \u001b[49m\u001b[43mdevice_ids\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m0,1\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m  \u001b[49m\n\u001b[32m     16\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     17\u001b[39m     summaries[(model_type, data_type, phase)] = summary\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Analysis/epiRNA/src/epiRNA/evals/benchmarks.py:64\u001b[39m, in \u001b[36mbenchmark_summary\u001b[39m\u001b[34m(model, dataloader, n_channels, loss_fn, count_pcc, count_gene, count_sum, device_ids)\u001b[39m\n\u001b[32m     61\u001b[39m y_pred, attns = model(x, rbp_expr, modality_idx)\n\u001b[32m     62\u001b[39m loss = loss_fn(y_pred, y)\n\u001b[32m---> \u001b[39m\u001b[32m64\u001b[39m total_loss += \u001b[43mloss\u001b[49m\u001b[43m.\u001b[49m\u001b[43mitem\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m * x.size(\u001b[32m0\u001b[39m)\n\u001b[32m     65\u001b[39m total_samples += x.size(\u001b[32m0\u001b[39m)\n\u001b[32m     67\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m count_pcc:\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: "
     ]
    }
   ],
   "source": [
    "summaries = {}\n",
    "for model_type, model in model_zoo.items():\n",
    "    # for data_type, datamodule in zip(\n",
    "    #     [\"complete\", \"maskrna\", \"maskseq\", \"maskrbp\"],\n",
    "    #     [fulldatamodule, maskrnadatamodule, maskseqdatamodule, maskrbpdatamodule]\n",
    "    # ):\n",
    "    if model_type == \"complete\":\n",
    "       datamodule = fulldatamodule\n",
    "    elif model_type == \"maskrna\":\n",
    "       datamodule = maskrnadatamodule\n",
    "    elif model_type == \"maskseq\":\n",
    "       datamodule = maskseqdatamodule\n",
    "    elif model_type == \"maskrbp\":\n",
    "       datamodule = maskrbpdatamodule\n",
    "    else:\n",
    "       raise ValueError(\"Unknown model type.\")\n",
    "        \n",
    "    dataloaders = [datamodule.train_dataloader(), datamodule.val_dataloader(), datamodule.test_dataloader()]\n",
    "    for phase, dataloader in zip([\"train\", \"val\", \"test\"], dataloaders):\n",
    "        print(f\"Evaluating Model: {model_type} on data during {phase} phase\")\n",
    "        summary = erna.evals.benchmark_summary(\n",
    "            model=model,\n",
    "            dataloader=dataloader,\n",
    "            n_channels=2,\n",
    "            data_device=\"cuda:2\",\n",
    "            device_ids=\"2,3\"  \n",
    "        )\n",
    "        summaries[(model_type, phase)] = summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{('complete', 'complete', 'train'): {'avg_loss': 0.014126870758198134,\n",
       "  'global_pcc': array([0.92665637, 0.        , 0.92665637], dtype=float32)},\n",
       " ('complete', 'complete', 'val'): {'avg_loss': 0.031025115417258088,\n",
       "  'global_pcc': array([0.765631, 0.      , 0.765631], dtype=float32)},\n",
       " ('complete', 'complete', 'test'): {'avg_loss': 0.03055985040441462,\n",
       "  'global_pcc': array([0.78595364, 0.        , 0.78595364], dtype=float32)},\n",
       " ('complete', 'maskrna', 'train'): {'avg_loss': 0.0714938507723525,\n",
       "  'global_pcc': array([0.67057407, 0.        , 0.67057407], dtype=float32)},\n",
       " ('complete', 'maskrna', 'val'): {'avg_loss': 0.06871554634139855,\n",
       "  'global_pcc': array([0.3864124, 0.       , 0.3864124], dtype=float32)},\n",
       " ('complete', 'maskrna', 'test'): {'avg_loss': 0.07447547057262828,\n",
       "  'global_pcc': array([0.39170244, 0.        , 0.39170244], dtype=float32)}}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "wrr_m6a_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
