{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import warnings\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "sys.path.append(\"/home/wangrr/Analysis/epiRNA/src\")\n",
    "import epiRNA as erna\n",
    "workdir = \"/data/wangrr/Analysis/epiRNA/workdir2/benchmark\"\n",
    "os.makedirs(workdir, exist_ok=True); os.chdir(workdir)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# model config:\n",
    "config = \"/home/wangrr/Analysis/epiRNA/src/experiments/model_v2/configs/basic_model.yaml\"\n",
    "\n",
    "# model checkpoint:\n",
    "ckpt_dir = \"/data/wangrr/Analysis/epiRNA/models\"\n",
    "checkpoints = {\n",
    "    \"human\": f\"{ckpt_dir}/v2human_only/version_0/checkpoints/epoch=13-step=76594-val_ePCC_mixed=0.7087.ckpt\",\n",
    "    \"mouse\": f\"{ckpt_dir}/v2mouse_only/version_0/checkpoints/epoch=08-step=52749-val_ePCC_mixed=0.7269.ckpt\"\n",
    "}\n",
    "\n",
    "# load models\n",
    "model_zoo = {}\n",
    "for key, ckpt in checkpoints.items():\n",
    "    model = erna.evals.load_model(ckpt, config)\n",
    "    model.eval()\n",
    "    model_zoo[key] = model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_version', 'gene_name', 'gene_source', 'gene_biotype', 'havana_gene', 'havana_gene_version', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'havana_transcript', 'havana_transcript_version', 'tag', 'transcript_support_level', 'exon_number', 'exon_id', 'exon_version', 'ccds_id', 'protein_id', 'protein_version']\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_version', 'gene_name', 'gene_source', 'gene_biotype', 'havana_gene', 'havana_gene_version', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'havana_transcript', 'havana_transcript_version', 'tag', 'transcript_support_level', 'exon_number', 'exon_id', 'exon_version', 'ccds_id', 'protein_id', 'protein_version']\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "\n",
    "datainfo = pd.read_excel(\"/home/wangrr/Analysis/epiRNA/data/mRNA_multiomic_sample_info.xlsx\", sheet_name=\"m6A\")\n",
    "\n",
    "ref_dir = \"/data/wangrr/Analysis/epiRNA/Dataset/Reference\"\n",
    "genome = {\n",
    "    \"human\":{\n",
    "        \"fasta\": f\"{ref_dir}/Homo_sapiens.GRCh38.dna.primary_assembly.84.fa\",\n",
    "        \"gtf\": f\"{ref_dir}/Homo_sapiens.GRCh38.84.gtf\"\n",
    "    },\n",
    "    \"mouse\":{\n",
    "        \"fasta\": f\"{ref_dir}//Mus_musculus.GRCm38.dna.primary_assembly.84.fa\",\n",
    "        \"gtf\": f\"{ref_dir}/Mus_musculus.GRCm38.84.gtf\"\n",
    "    }\n",
    "}\n",
    "\n",
    "human_gtf_df = erna.pp.GTFLoader(gtf_path=genome[\"human\"][\"gtf\"], zero_based=True)\n",
    "mouse_gtf_df = erna.pp.GTFLoader(gtf_path=genome[\"mouse\"][\"gtf\"], zero_based=True)\n",
    "\n",
    "chrom_kwargs = {\n",
    "    'human': {\n",
    "        'train_chroms': [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"9\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\", \"20\", \"21\", \"22\"],\n",
    "        'val_chroms': [\"8\"],\n",
    "        'test_chroms': [\"10\"]\n",
    "    },\n",
    "    'mouse': {\n",
    "        'train_chroms': [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"9\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\"],\n",
    "        'val_chroms': [\"8\"],\n",
    "        'test_chroms': [\"10\"]\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_samples = [\"Adipose-1\", \"Aorta-1\", \"Appendix-1\", \"Esophagus-1\", \"GOS-1\", \"Hela-1\", \"HT29-1\", \"Hypothalamus-1\", \"Jejunum-1\", \"Liver-1\", \"Prostate-1\", \"Skin-1\", \"Testis-1\", \"Tongue-1\", \"Trachea-1\", \"WPMY-1\", \"MT4-1\", \"GSC11-1\", \"iSLK-1\", \"TIME-1\"]\n",
    "mouse_samples = [\"mHeart-1\", \"mSpleen-1\", \"mLung-1\", \"mLiver-1\", \"mCerebrum-1\", \"mCerebellum-1\", \"mBrainstem-1\", \"mHypothalamus-1\", \"mBMDC-1\", \"mEF-1\", \"mNPC-1\", \"mESC-1\", \"mStriatum-1\", \"mPeritoneal_macrophages-1\", \"mB16-OVA-1\", \"miPSC-1\", \"mNSC-1\", \"mForebrain-1\", \"mKidney-1\", \"mHypothalamus-2\"]\n",
    "\n",
    "human_datainfo = datainfo[datainfo[\"SID\"].isin(human_samples)].reset_index(drop=True).copy()\n",
    "mouse_datainfo = datainfo[datainfo[\"SID\"].isin(mouse_samples)].reset_index(drop=True).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"/home/wangrr/Analysis/epiRNA/data/rbp_expression_dict.pkl\", \"rb\") as f:\n",
    "    rbp_dict = pickle.load(f)['rbp_dict']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total samples: 350180\n",
      "Omics types: ['m6A']\n",
      "Total samples: 13380\n",
      "Omics types: ['m6A']\n",
      "Total samples: 14580\n",
      "Omics types: ['m6A']\n",
      "Total samples: 375140\n",
      "Omics types: ['m6A']\n",
      "Total samples: 21040\n",
      "Omics types: ['m6A']\n",
      "Total samples: 20480\n",
      "Omics types: ['m6A']\n"
     ]
    }
   ],
   "source": [
    "human_datasetkwargs = {\n",
    "    'human_fasta_path': genome[\"human\"][\"fasta\"],\n",
    "    'mouse_fasta_path': genome[\"mouse\"][\"fasta\"],\n",
    "    \"modality_to_index\": {\"m6A_human\": 0, \"m6A_mouse\": 1},\n",
    "    'bigwig_dir': \"/data/wangrr/Analysis/epiRNA/Dataset/processed/4-coverage\",\n",
    "    'datainfo': human_datainfo,\n",
    "    'rbp_dict': rbp_dict,\n",
    "    'human_gtf_df': human_gtf_df,\n",
    "    'mouse_gtf_df': mouse_gtf_df,\n",
    "    'return_bin_input': True,\n",
    "    'seq_len': 65536,\n",
    "    'mask_type': None\n",
    "}\n",
    "\n",
    "mouse_datasetkwargs = human_datasetkwargs.copy()\n",
    "mouse_datasetkwargs.update({\n",
    "    'datainfo': mouse_datainfo\n",
    "})\n",
    "\n",
    "human_datamodule = erna.ds.GeneDataModule(\n",
    "    dataset_cls=erna.ds.GeneDataset,\n",
    "    dataset_kwargs=human_datasetkwargs,\n",
    "    chrom_kwargs=chrom_kwargs,\n",
    "    batch_size=32,\n",
    "    num_workers=8\n",
    ")\n",
    "\n",
    "mouse_datamodule = erna.ds.GeneDataModule(\n",
    "    dataset_cls=erna.ds.GeneDataset,\n",
    "    dataset_kwargs=mouse_datasetkwargs,\n",
    "    chrom_kwargs=chrom_kwargs,\n",
    "    batch_size=32,\n",
    "    num_workers=8\n",
    ")\n",
    "\n",
    "human_datamodule.setup(stage='test')\n",
    "mouse_datamodule.setup(stage='test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating Model: mouse on data during val phase\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Benchmarking PCC:   0%|          | 0/658 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Benchmarking PCC: 100%|██████████| 658/658 [02:30<00:00,  4.38it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'avg_loss': 0.050114666274483154, 'global_pcc': array([0.       , 0.7264283, 0.7264283], dtype=float32)}\n",
      "Evaluating Model: mouse on data during train phase\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Benchmarking PCC: 100%|██████████| 11723/11723 [44:17<00:00,  4.41it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'avg_loss': 0.02222799146243643, 'global_pcc': array([0.       , 0.8568593, 0.8568593], dtype=float32)}\n",
      "Evaluating Model: mouse on data during test phase\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Benchmarking PCC: 100%|██████████| 640/640 [02:25<00:00,  4.39it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'avg_loss': 0.043096711458019854, 'global_pcc': array([0.       , 0.7361933, 0.7361933], dtype=float32)}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "summaries = {}\n",
    "for model_type, model in model_zoo.items():\n",
    "    if model_type == \"human\":\n",
    "        datamodule = human_datamodule\n",
    "    elif model_type == \"mouse\":\n",
    "        datamodule = mouse_datamodule\n",
    "    else:\n",
    "        continue\n",
    "    dataloaders = [datamodule.val_dataloader(), datamodule.train_dataloader(), datamodule.test_dataloader()]\n",
    "    for phase, dataloader in zip([\"val\", \"train\", \"test\"], dataloaders):\n",
    "        print(f\"Evaluating Model: {model_type} on data during {phase} phase\")\n",
    "        summary = erna.evals.benchmark_summary(\n",
    "            model=model,\n",
    "            dataloader=dataloader,\n",
    "            n_channels=2,\n",
    "            data_device=\"cuda:0\",\n",
    "            device_ids=\"0,1\"  \n",
    "        )\n",
    "        summaries[(model_type, phase)] = summary\n",
    "        print(summary)\n",
    "            \n",
    "with open(\"benchmark_species_summaries.pkl\", \"wb\") as f:\n",
    "    pickle.dump(summaries, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "wrr_m6a_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
