{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pickle\n",
    "import numpy as np\n",
    "import torch\n",
    "import pandas as pd\n",
    "sys.path.append(\"/home/wangrr/Analysis/epiRNA/src\")\n",
    "import epiRNA as erna\n",
    "\n",
    "workdir = \"/data/wangrr/Analysis/epiRNA/workdir2/attribution\"\n",
    "os.makedirs(workdir, exist_ok=True); os.chdir(workdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "checkpoints = \"/data/wangrr/Analysis/epiRNA/models/basic_fold/version_0/checkpoints/epoch=18-step=215327-val_ePCC_mixed=0.7823.ckpt\"\n",
    "config_file = \"/home/wangrr/Analysis/epiRNA/src/experiments/model_v2/configs/basic_model.yaml\"\n",
    "model = erna.evals.load_model(ckpt=checkpoints, config_yaml=config_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_version', 'gene_name', 'gene_source', 'gene_biotype', 'havana_gene', 'havana_gene_version', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'havana_transcript', 'havana_transcript_version', 'tag', 'transcript_support_level', 'exon_number', 'exon_id', 'exon_version', 'ccds_id', 'protein_id', 'protein_version']\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_version', 'gene_name', 'gene_source', 'gene_biotype', 'havana_gene', 'havana_gene_version', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'havana_transcript', 'havana_transcript_version', 'tag', 'transcript_support_level', 'exon_number', 'exon_id', 'exon_version', 'ccds_id', 'protein_id', 'protein_version']\n"
     ]
    }
   ],
   "source": [
    "datainfo = pd.read_excel(\"/home/wangrr/Analysis/epiRNA/data/mRNA_multiomic_sample_info.xlsx\", sheet_name=\"m6A\")\n",
    "\n",
    "samples = [\"Cerebrum-1\", \"Cerebellum-1\"]\n",
    "datainfo = datainfo[datainfo[\"SID\"].isin(samples)].reset_index(drop=True)\n",
    "\n",
    "ref_dir = \"/data/wangrr/Analysis/epiRNA/Dataset/Reference\"\n",
    "genome = {\n",
    "    \"human\":{\n",
    "        \"fasta\": f\"{ref_dir}/Homo_sapiens.GRCh38.dna.primary_assembly.84.fa\",\n",
    "        \"gtf\": f\"{ref_dir}/Homo_sapiens.GRCh38.84.gtf\"\n",
    "    },\n",
    "    \"mouse\":{\n",
    "        \"fasta\": f\"{ref_dir}//Mus_musculus.GRCm38.dna.primary_assembly.84.fa\",\n",
    "        \"gtf\": f\"{ref_dir}/Mus_musculus.GRCm38.84.gtf\"\n",
    "    }\n",
    "}\n",
    "\n",
    "human_gtf_df = erna.pp.GTFLoader(gtf_path=genome[\"human\"][\"gtf\"], zero_based=True)\n",
    "mouse_gtf_df = erna.pp.GTFLoader(gtf_path=genome[\"mouse\"][\"gtf\"], zero_based=True)\n",
    "\n",
    "chrom_kwargs = {\n",
    "    'human': {\n",
    "        'train_chroms': [],\n",
    "        'val_chroms': [str(i) for i in range(1, 23)],\n",
    "        'test_chroms': []\n",
    "    },\n",
    "    'mouse': {\n",
    "        'train_chroms': [],\n",
    "        'val_chroms': [],\n",
    "        'test_chroms': []\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total samples: 0\n",
      "Omics types: ['m6A']\n",
      "Total samples: 37814\n",
      "Omics types: ['m6A']\n",
      "Total samples: 0\n",
      "Omics types: ['m6A']\n"
     ]
    }
   ],
   "source": [
    "rbp_dict_file = \"/home/wangrr/Analysis/epiRNA/data/rbp_expression_dict.pkl\"\n",
    "with open(rbp_dict_file, \"rb\") as f:\n",
    "    rbp_dict = pickle.load(f)\n",
    "    \n",
    "dataset_kwargs = {\n",
    "    'human_fasta_path': genome[\"human\"][\"fasta\"],\n",
    "    'mouse_fasta_path': genome[\"mouse\"][\"fasta\"],\n",
    "    \"modality_to_index\": {\"m6A_human\": 0, \"m6A_mouse\": 1},\n",
    "    'bigwig_dir': \"/data/wangrr/Analysis/epiRNA/Dataset/processed/4-coverage\",\n",
    "    'datainfo': datainfo,\n",
    "    'rbp_dict': rbp_dict['rbp_dict'],\n",
    "    'human_gtf_df': human_gtf_df,\n",
    "    'mouse_gtf_df': mouse_gtf_df,\n",
    "    'return_bin_input': True,\n",
    "    'seq_len': 65536,\n",
    "    'mask_type': None\n",
    "}\n",
    "\n",
    "datamodule = erna.ds.GeneDataModule(\n",
    "    dataset_cls=erna.ds.GeneDataset,\n",
    "    dataset_kwargs=dataset_kwargs,\n",
    "    chrom_kwargs=chrom_kwargs,\n",
    "    batch_size=32,\n",
    "    num_workers=8\n",
    ")\n",
    "datamodule.setup(stage=\"test\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets = datamodule.val_dataset\n",
    "dataloader = datamodule.val_dataloader()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/1182 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1182/1182 [30:43<00:00,  1.56s/it]\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "ig_score = []\n",
    "igattributor = erna.ex.attribution.InputGradientAttrbution(model=model.to(device=\"cuda:0\"))\n",
    "target_manipulator = erna.ex.attribution.Manipulation(method=\"sum\")\n",
    "for batch in tqdm(dataloader):\n",
    "    x = batch[\"input\"].to(dtype=torch.float32, device=\"cuda:0\")\n",
    "    y = batch[\"target\"].to(dtype=torch.float32, device=\"cuda:0\")\n",
    "    rbp_expr = batch['rbp_expr'].to(dtype=torch.float32, device=\"cuda:0\")\n",
    "    modality_idx = batch[\"modality_idx\"].to(dtype=torch.int64, device=\"cuda:0\")\n",
    "    ig_scores_batch = igattributor.attribute(\n",
    "        inputs=x, \n",
    "        target_manipulation=target_manipulator,\n",
    "        additional_forward_args=(rbp_expr, modality_idx),\n",
    "    )\n",
    "    ig_score.append(ig_scores_batch)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# concatenate all batches\n",
    "ig_score = torch.cat(ig_score, dim=0)\n",
    "\n",
    "# get sequence attribution scores only\n",
    "seq_ig_score = ig_score[:, :4, :].cpu().numpy()  #\n",
    "\n",
    "# get expression attribution scores only\n",
    "expr_ig_score = ig_score[:, 4:, :].cpu().numpy()  #\n",
    "\n",
    "# get sum attribution scores over the ACGU channel\n",
    "sum_seq_ig_score = np.sum(seq_ig_score, axis=1)  # shape: (num_samples, seq_len)\n",
    "\n",
    "# save all\n",
    "np.savez_compressed(\n",
    "    \"cns_m6A_ig_attribution.npz\",\n",
    "    seq_ig_score=seq_ig_score,\n",
    "    expr_ig_score=expr_ig_score,\n",
    "    sum_seq_ig_score=sum_seq_ig_score\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "wrr_m6a_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
